va/caa.go

```diff
@@ -40,16 +49,164 @@ func (va *ValidationAuthorityImpl) IsCAAValid(ctx context.Context, req *vapb.IsC
+		}).Observe(localRecheckLatency.Seconds())
+		if prob == nil {
+			va.metrics.caaRecheckTime.With(prometheus.Labels{
+				"type":         req.ValidationMethod,
+				"result":       checkResult,
+				"problem_type": "",
+			}).Observe(recheckLatency.Seconds())
+		} else {
+			va.metrics.caaRecheckTime.With(prometheus.Labels{
+				"type":         req.ValidationMethod,
+				"result":       checkResult,
+				"problem_type": string(prob.Type),
+			}).Observe(recheckLatency.Seconds())
+		}
+	}
+
+	if prob != nil {
+		return &vapb.IsCAAValidResponse{Problem: pbProblem}, nil
+	} else {
+		return &vapb.IsCAAValidResponse{}, nil
+	}
+}
+
+// processRemoteCAARecheckResultsOuter is a helper method that supplies metadata
+// to the generic `va.processRemoteResultsInner` method and returns its problem
+// details.
+func (va *ValidationAuthorityImpl) processRemoteCAARecheckResultsOuter(domain string, acctID int64, challengeType string, primaryResult *probs.ProblemDetails, remoteResultsChan chan *remoteVAResult) *probs.ProblemDetails {
+	var isOk bool
+	start := va.clk.Now()
+
+	defer func(ok *bool) {
+		var result string
+		if *ok {
+			result = "success"
+		} else {
+			result = "failure"
+		}
+
+		va.metrics.remoteCAARecheckTime.With(prometheus.Labels{
+			"type":   challengeType,
+			"result": result,
+		}).Observe(va.clk.Since(start).Seconds())
+	}(&isOk)
+
+	metadata := remoteMetadata{
+		rpc:    "IsCAAValid",
+		action: "CAA rechecking",
+	}
+	probs, isOk := va.processRemoteResultsInner(domain, acctID, challengeType, primaryResult, remoteResultsChan, metadata)
+
+	return probs
+}
+
+// performRemoteCAARecheck calls `isCAAValid` for each of the configured
+// remoteVAs in a random order. The provided `results` chan should have an equal
+// size to the number of remote VAs. The CAA rechecks will be performed in
+// separate go-routines. If the result `error` from a remote `isCAAValid` RPC is
+// nil or a nil `ProblemDetails` instance it is written directly to the
+// `results` chan. If the err is a cancelled error it is treated as a nil error.
+// Otherwise the error/problem is written to the results channel as-is.
+func (va *ValidationAuthorityImpl) performRemoteCAARecheck(
+	ctx context.Context,
+	req *vapb.IsCAAValidRequest,
+	results chan *remoteVAResult) {
+	for _, i := range rand.Perm(len(va.remoteVAs)) {
+		remoteVA := va.remoteVAs[i]
+		go func(rva RemoteVA) {
+			result := &remoteVAResult{
+				VAHostname: rva.Address,
+			}
+			// Recursively call IsCAAValid, but with a remoteVA masquerading as
+			// a fully-fledged VA.
+			res, err := rva.IsCAAValid(ctx, req)
+			if err != nil && canceled.Is(err) {
+				// If the non-nil err was a canceled error, ignore it. That's fine: it
+				// just means we cancelled the remote VA request before it was
+				// finished because we didn't care about its result. Don't log to avoid
+				// spamming the logs.
+				result.Problem = probs.ServerInternal("Remote IsCAAValidRequest RPC canceled")
```

Comment:
Here and in the five other messages below:
```suggestion
				result.Problem = probs.ServerInternal("Remote IsCAAValid RPC canceled")
```